### PR TITLE
move building image to gcloud container build and begin migrating to new test framework

### PR DESCRIPTION
**Pull request by nkubala**
This PR represents the first step towards unifying the CI/CD process for the silver language runtime images. It transfers the build process for constructing the image to a cloud container build, and integrates a [structure test build step] (https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/structure_tests) to run basic tests on the built image before pushing.

Additionally, another build step is added for running `rake` in order to support the legacy integration tests. This comes with the bonus of again exiting the build if any of the integration tests fail, since they are run before the final step of the cloud build. 

Since these integration tests are containerized sample apps, I created a simple image that is a bare-bones container with docker and rake installed, and pushed it to gcr.io/gcp-runtimes (though it can live whereever). It should be noted that eventually, there will be an integration testing framework (similar to the structure testing framework) which will further generalize these tests.

Also worth noting; this build currently takes ~15 minutes. The default timeout for `gcloud alpha container builds create` is 10 minutes, so I had to update the property on my account manually to get it to succeed. Not sure if there's a way to embed this in the build config.

@hxiong388 @dazuma @dlorenc

### CODE REVIEW

File: appengine/test/tc_sample_apps.rb [hunk 1]
```diff
@@ -32,7 +32,7 @@ def run_app_test(dirname)
     ::Dir.chdir(::File.join(APPS_DIR, dirname)) do |dir|
       build_docker_image("", dirname) do |image|
         run_docker_daemon("-p 8080:8080 #{image}") do |container|
-          assert_cmd_output("curl -s -S http://127.0.0.1:8080/", "ruby app", 10)
+          assert_cmd_output("docker run --link #{container}:#{container} gcr.io/google-appengine/debian8 /bin/bash -c 'apt-get -qq update > /dev/null 2>&1; apt-get -qq -y install curl > /dev/null 2>&1; curl -s -S #{container}:8080/'", "ruby app", 15)
```
Comment (dlorenc) [inline appengine/test/tc_sample_apps.rb]: Ooh, this is a little annoying, did you try to run this in the ruby image itself instead? That way we could avoid linking and having to install curl.

File: rake_image/cloudbuild.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+steps:
+  - name: gcr.io/cloud-builders/docker
+    args: ['build', 't', 'gcr.io/gcp-runtimes/rake', '.']
```
Comment (dlorenc) [inline rake_image/cloudbuild.yaml]: Does this need to be -t?

File: appengine/test/tc_sample_apps.rb (hunk 1 above)
Comment (nkubala) [inline appengine/test/tc_sample_apps.rb]: wow that's obviously the better solution haha. shaves a minute off of the build time too. good call.

File: rake_image/cloudbuild.yaml (hunk 2 above)
Comment (nkubala) [inline rake_image/cloudbuild.yaml]: hmm....yeah that's definitely what it should be, i'm wondering how this worked in the first place.

File: appengine/test/tc_sample_apps.rb [hunk 3]
```diff
@@ -32,7 +32,7 @@ def run_app_test(dirname)
     ::Dir.chdir(::File.join(APPS_DIR, dirname)) do |dir|
       build_docker_image("", dirname) do |image|
         run_docker_daemon("-p 8080:8080 #{image}") do |container|
-          assert_cmd_output("docker run --link #{container}:#{container} gcr.io/google-appengine/debian8 /bin/bash -c 'apt-get -qq update > /dev/null 2>&1; apt-get -qq -y install curl > /dev/null 2>&1; curl -s -S #{container}:8080/'", "ruby app", 15)
+          assert_cmd_output("docker exec #{container} curl -s -S http://127.0.0.1:8080", "ruby app", 10)
```
Comment (hxiong388) [inline appengine/test/tc_sample_apps.rb]: These tests are all failing on me due to this same error:
```sh
Step #2: docker exec ruby-test-container-generic-e1971c00 curl -s -S http://127.0.0.1:8080
Step #2: curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
```

File: appengine/test/tc_sample_apps.rb [hunk 4]
```diff
@@ -32,7 +32,7 @@ def run_app_test(dirname)
     ::Dir.chdir(::File.join(APPS_DIR, dirname)) do |dir|
       build_docker_image("", dirname) do |image|
         run_docker_daemon("-p 8080:8080 #{image}") do |container|
-          assert_cmd_output("curl -s -S http://127.0.0.1:8080/", "ruby app", 10)
+          assert_cmd_output("docker exec #{container} curl -s -S http://127.0.0.1:8080", "ruby app", 10)
```
Comment (hxiong388) [inline appengine/test/tc_sample_apps.rb]: These tests are all failing on me due to this same error:
```sh
Step #2: docker exec ruby-test-container-generic-e1971c00 curl -s -S http://127.0.0.1:8080
Step #2: curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
```

Comment (nkubala) [inline appengine/test/tc_sample_apps.rb]: do the tests actually fail for you? when I run them, it usually takes a few tries before the container is fully up and serving requests, but eventually the client gets a response.

```
Successfully built 801de2b1f753
docker run --name ruby-test-container-generic-c4331d1f -d -p 8080:8080 ruby-test-image-rails4_app-e7aece60
367703c1f169f81e84c00e17775b573df60a8dcd5bee5632e50a714835a70e18
docker exec ruby-test-container-generic-c4331d1f curl -s -S http://127.0.0.1:8080
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
...expected result did not arrive yet (iteration 0)...
docker exec ruby-test-container-generic-c4331d1f curl -s -S http://127.0.0.1:8080
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
...expected result did not arrive yet (iteration 1)...
docker exec ruby-test-container-generic-c4331d1f curl -s -S http://127.0.0.1:8080
docker kill ruby-test-container-generic-c4331d1f
ruby-test-container-generic-c4331d1f
docker rm ruby-test-container-generic-c4331d1f
ruby-test-container-generic-c4331d1f
docker rmi ruby-test-image-rails4_app-e7aece60
Untagged: ruby-test-image-rails4_app-e7aece60:latest
```

Comment (hxiong388) [inline appengine/test/tc_sample_apps.rb]: Ok, I see. I run the cloud build script multiple times and it always comes back with "FAILURE" status, so I automatically assumed it was the refused connection error from the tests.

After a closer look as you pointed out, I realize the connection takes multiple tries, and my failure build may be result from the final image push isn't denied.

**Pull request merged**